File: project_euler/src/problem_052.erl
%% It can be seen that the number, 125874, and its double, 251748, contain exactly the same digits, but in a different order.
%% Find the smallest positive integer, x, such that 2x, 3x, 4x, 5x, and 6x, contain the same digits.

-module(problem_052).

-export([get_check_data/0, prepare_data/2, solve/1]).

-behaviour(numerical_task_behaviour).

get_check_data() ->
    [{none, 142857}].

prepare_data(_ModuleSourceDir, Input) -> Input.

solve(none) -> search(1).

search(DigitCount) ->
    MinNumber = numbers:power(10, DigitCount - 1),
    MaxNumber = (numbers:power(10, DigitCount) - 1) div 6,
    case search(MinNumber, MaxNumber) of
        {true, SearchedNumber} -> SearchedNumber;
        false -> search(DigitCount - 1)
    end.

search(CurrentNumber, MaxNumber) when MaxNumber < CurrentNumber -> false;
search(CurrentNumber, MaxNumber) ->
    case check_number(CurrentNumber, get_result_numbers(CurrentNumber)) of
        true -> {true, CurrentNumber};
        false -> search(MaxNumber, CurrentNumber + 1)
    end.

check_number(Number, ExpectedResult) ->
    NumberDigits = lists:sort(numbers:get_digits(Number)),
    lists:all(fun(ExpectedNumber) -> NumberDigits == lists:sort(ExpectedNumber) end, ExpectedResult).

get_result_numbers(Number) ->
    [numbers:get_digits(Number * Factor) || Factor <- lists:seq(2, 6)].

File: projecteuler.fsharp/ProjectEulerTasks/Problem052.fs
namespace ProjectEulerTasks

open NUnit.Framework
open ProjectEulerTasks.Utils
open CommonLib

// It can be seen that the number, 125874, and its double, 251748, contain exactly the same digits, but in a different order.
// Find the smallest positive integer, x, such that 2x, 3x, 4x, 5x, and 6x, contain the same digits.

[<TestFixture>]
type Problem052() =

    let checkNumber (number: int) =
        let sourceDigits = number |> NumbersDigits.GetDigits |> List.sort
        seq {2 .. 6} |> Seq.exists (fun factor -> factor * number |> NumbersDigits.GetDigits |> List.sort = sourceDigits |> not) |> not

    let rec searchNumber (digitsCount: int) =
        let minNumber = pown 10 (digitsCount - 1)
        let maxNumber = ((pown 10 digitsCount) - 1) / 6
        match seq {minNumber .. maxNumber} |> Seq.tryFind checkNumber with
        | None -> searchNumber (digitsCount + 1)
        | Some number -> number

    let solveImpl () = searchNumber 1

    [<TestCase(142857, TimeThresholds.HardTimeLimit)>]
    member public this.Solve(expectedAnswer: int, timeLimit: int) =
        SolutionUtils.CheckSolution(timeLimit, expectedAnswer, solveImpl)